keps/sig-auth/0034-20190123-harden-default-discovery-bindings.md

```diff
@@ -0,0 +1,71 @@
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+The aim of this change is to remove the `system:unauthenticated` subjects from the `system:discovery` and `system:basic-user` ClusterRoleBindings, while preserving unauthenticated access to a genuinely non-sensitive subset of the current APIs (e.g. `GET /healthz`, `GET /version`). This will improve the default privacy and security posture for new clusters.
+
+## Motivation
+
+One of the work items resulting from the [CVE-2018-1002105](https://github.com/kubernetes/kubernetes/issues/71411) [post-mortem](https://docs.google.com/document/d/1tGpqEM56cSBPfl-qxIs_CGdlABMZXEKEDhCI7eOwGT0/preview) was to [investigate hardening the default RBAC discovery ClusterRoleBindings](https://github.com/kubernetes/kubernetes/issues/72115) (i.e. `system:discovery` and `system:basic-user`) to limit potential avenues for similar attack. Additionally, the fact that API extensions are exposed by the default discovery bindings is surprising to some and represents a potential privacy concern (e.g. `/apis/self-flying-cars.lyft.io/v1/`).
+
+### Goals
+
+Limit the set of APIs which allow for unauthenticated access by default.
+
+### Non-Goals
+
+Disabling unauthenticated and/or anonymous access to all APIs by default.
+
+## Proposal
+
+* Remove the `GET /healthz`, `/version` and `/version/` PolicyRules from the default [`system:discovery` ClusterRole](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L198-L208)
+* Remove the `system:unauthenticated` subject group from the default [`system:discovery` and `system:basic-user` ClusterRoleBindings](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L526-L527)
+* Create a new default ClusterRole which prominently, explicitly documents its intent to freely disclose information (e.g. `system:public-info-viewer`) and contains the `GET /healthz`, `/version` and `/version/` PolicyRules removed from `system:discovery`
+* Create a new default ClusterRoleBinding for the new ClusterRole, which grants the `system:(un)authenticated` subject groups access.
+
+### Risks and Mitigations
+
+1. If a user upgrades a cluster which has already limited access to the `GET /healthz`, `/version` or `/version/` APIs, auto-reconciliation will re-enable universal access via `system:public-info-viewer`
+  1. We'll explicitly call this out as an `ACTION REQUIRED` with a sample `kubectl apply` which can be used to effectively disable the `system:public-info-viewer` ClusterRoleBinding
+1. Some use-cases might require the existing permissions to be preserved for unauthenticated calls
+  1. In the release notes, we can include an easy 'escape hatch' to re-enable unauthenticated access to the APIs, e.g. `kubectl create clusterrolebinding public-discovery --clusterrole=system:discovery --group=system:unauthenticated`
```

Comment:
This might need to follow the deprecation policy, and roll out over multiple releases.

Reply:
Guidance accepted...

Reply:
Since the proposed approach lets upgraded clusters work as-is without any behavior change, I think this can be implemented immediately.